### ISSUE TITLE
2018 Spring Cleaning

### ISSUE THREAD

**Issue by cmyr** [xi-editor/xi-editor#558]
_tl;dr: a very loosely structured overview of a bunch of mostly non-functional changes planned for the next little while._

# 2018 Spring Cleaning

In anticipation of GSoC and other upcoming work, I would like to do some refactoring over the next few months. Much of this is long planned/previously discussed; I'll link to existing issues where possible.

I'm going to take a fairly high-level view, here; some of these items will need to be fleshed out in more detail in subsequent issues.

## First (preliminary architecture changes)

There are a number of changes I would like to land in the near future, which will pave the way for future work:
  
1. _Global plugins_. (#472) Currently (in the general case) a plugin is bound to a particular view; multiple views which want to use the same plugin each launch their own instance. I would like to make plugins global-by-default, where a single plugin process is expected to handle multiple views.on

2. _Buffer tags and plugin lifespans_. I would like to essentially adopt the model proposed by @trishume in #374. Buffers will be assigned 'tags', which will supercede fixed syntax definitions. Plugins can provide new tags; plugins and the frontend can add/remove tags; plugins can declare the tags that they should be activated for.

2. (2b) _Remove support for non-global plugins?_. This needs a bit more discussion, but I it is unclear to me that the possible advantages of non-global plugins (where each buffer has its own running process) outweigh the increase in API complexity.

## Second (RPC refactor)

At some later point, I would also like to do some work on the RPC library, with the goal of adding support for async I/O where possible, and for handling multiple peers in a single runloop.

### How it works now

Currently, each plugin has its own instance of the runloop, running in a thread; since each runloop starts a distinct thread to handle reading, that means each plugin is currently starting two threads _in addition_ to the plugin process itself.

In addition, since each plugin is running in its own thread, access to buffers has to go through a mutex; this makes a lot of the API around interacting with buffers more complicated than it needs to be.

### Goals

I don't have a coherent design right now, but i do have an outline of what I would like to achieve:

- all connections to peers (client, plugins, possibly others) should be handled by a single runloop.

- where possible, connections to peers should use async APIs under the hood; where that isn't possible, we can use a blocking read thread as a shim.

- All messages should be handled on a single main thread; this removes a bunch of complexity around sharing buffers between threads.

- we should be able to add and remove peers during execution, by passing a handle to the runloop

- the runloop should support timers, so we can do things like wait momentarily for plugin actions to arrive before updating the client

- the runloop should be easy to deploy as a library, and provide FFI bindings

- the runloop should be agnostic in terms of peer implementation; it should be able to support connecting to peers over stdio, sockets, channels (peers that are local, but running in their own thread) and whatever other transport mechanism we might desire.

### Implementation questions

It would be good to rely on existing work for this, where possible. To that end, I'm mostly waiting for futures/tokio 0.2 to crystalize; it would be nice to lean on an existing runloop implementation as much as possible. That said, It isn't totally clear ot me that futures are something we need in and of themselves. While futures would be a nice abstraction for certain kinds of operations, they aren't (to the best of my understanding) particularly helpful in the general case, where we're just sending and receiving fully asynchronous messages.

## Third (unification)

Once these two bits of work are done, it will be possible to generally rationalize how we recieve and route RPCs, and to simplify how we update peers after performing an action.

This is tentative; there are a few details here that I'll need to play around with more before I feel confident about a concrete implementation.

1. _View centrism_ (#371). When an edit RPC arrives it is associated with a view identifier; we then have to do a little dance to go from a view identifier to a _buffer_ identifier, which we use to retreive the actual editor that will handle the RPC. This involves unwraps at various stages, which are bad, because it's entirely possible for an editor to stop existing between when an RPC is sent and it is to be handled. I would propose instead routing RPCs through views; views would hold a reference to their associated editor (either through an Rc or an identifier) and would be passed to the editor as an argument.

2. _Central dispatch_ (very maybe) all edit actions from various sources arrive centrally; they are applied to the relevant resources (buffers, views) and then deltas are dispatched to interested parties, all without any tricky synchronization code. This is too distant for me comment on in too much detail, but I think we would be in a good position to remove a lot of complexity, and make future feature development easier.

**Comment by cmyr** [comment]
Closing this, although certain things (the RPC refactor, plugin lifespan) remain unresolved.
